DAOS-6174 control: Adjust pool create retry behavior

mjmac opened the pull request:
Moves the logic for retrying pool creation operations to
the client side of the Control API. In hindsight,
this should have been the approach all along because it will
work better with multiple MS replicas. Makes a change to
the target creation logic to intercept a -DER_EXISTS and
convert it to -DER_AGAIN to signal that a pool creation or
cleanup process is already in progress and that the
control plane should retry later.

Also adds support for request deadlines to better control client
behavior. Bumps default pool create timeout to 10 minutes to
accommodate very large pools. Longer-term, we should set this
timeout based on SCM size when we understand the relationship
better.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3997/1/execution/node/57/log

mjmac commented:
@liw: Added you for comment on the change to per-try pool UUIDs. Overall I think that we should move the retry logic back out to the client to work better with multiple MS replicas, but I'm wondering whether it would be a mistake to make each retry have its own pool UUID. I can see a case being made for one user request being tied to one UUID, in order to avoid accidentally creating multiple pools due to some error in the retry logic.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3997/2/execution/node/58/log

kjacque commented on src/control/lib/control/request.go:
```diff
@@ -80,12 +98,45 @@ func (r *request) AddHost(hostAddr string) {
+}
+
+// onRetry implements the retryer interface and always returns
+// an error. Callers should check the result of canRetry() before
+// calling retry, in order to avoid wasting effort on a request
+// that does not implement its own retry logic.
```
This part of the comment seems to be more about the interface in general than about this specific implementation. I think it'd be helpful to put any general comments about how the interface aught to be used near the interface definition.

mjmac replied:
fixed

kjacque commented on src/control/lib/control/request.go:
```diff
@@ -97,14 +148,58 @@ func (r *msRequest) isMSRequest() bool {
+}
+
+// retryAfter returns a channel to be waited on in order to pause
+// a goroutine's execution for some duration. For safety, the
+// channel should be waited on in a select with a context in order to
+// avoid blocking a goroutine forever.
```
Another comment that seems a bit more general, about how callers should behave with the interface.

mjmac replied:
fixed

kjacque reviewed (approved):
LGTM. Just a couple minor comments about comments.

liw commented:
> ~@liw: Added you for comment on [...] associating all retries with the same pool UUID. This seems cleanest and least likely to result in ambiguous behavior.

Agreed. I've left a comment on DAOS-6174.

daosbuild1 commented on src/control/lib/control/request.go:
```diff
@@ -80,12 +109,45 @@ func (r *request) AddHost(hostAddr string) {
 }
 
+// SetTimeout sets a deadline by which the request must have
+// completed. It is calcluated from the supplied time.Duration.
```
```suggestion
// completed. It is calculated from the supplied time.Duration.
```

daosbuild1 commented on src/control/lib/control/request.go:
```diff
@@ -80,12 +109,45 @@ func (r *request) AddHost(hostAddr string) {
+
+// retryAfter implements the retrier interface and always returns
+// a time channel defined by the supplied duration.
+func (r *request) retryAfter(dur time.Duration) <-chan time.Time {
+	return time.After(dur)
```
```suggestion
func (r *request) retryAfter(due time.Duration) <-chan time.Time {
	return time.After(due)
```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3997/3/execution/node/59/log

daosbuild1 commented:
Test stage NLT completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3997/3/display/redirect

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mjmac commented:
Passed build and unit tests. No need to re-run functional tests as the changes since the last full run were to documentation and to fix a codespell complaint.
